src/bin/pgcopydb/sentinel.c

```diff
@@ -0,0 +1,582 @@
+	char writeLSN[PG_LSN_MAXLENGTH] = { 0 };
+	char flushLSN[PG_LSN_MAXLENGTH] = { 0 };
+
+	sformat(startLSN, sizeof(startLSN), "%X/%X", LSN_FORMAT_ARGS(flush_lsn));
+	sformat(writeLSN, sizeof(writeLSN), "%X/%X", LSN_FORMAT_ARGS(write_lsn));
+	sformat(flushLSN, sizeof(flushLSN), "%X/%X", LSN_FORMAT_ARGS(flush_lsn));
+
+	/* bind our parameters now */
+	BindParam params[] = {
+		{ BIND_PARAMETER_TYPE_TEXT, "startpos", 0, (char *) startLSN },
+		{ BIND_PARAMETER_TYPE_TEXT, "write_lsn", 0, (char *) writeLSN },
+		{ BIND_PARAMETER_TYPE_TEXT, "flush_lsn", 0, (char *) flushLSN }
+	};
+
+	int count = sizeof(params) / sizeof(params[0]);
+
+	if (!catalog_sql_bind(&query, params, count))
+	{
+		/* errors have already been logged */
+		(void) semaphore_unlock(&(catalog->sema));
+		return false;
+	}
+
+	/* now execute the query, which does not return any row */
+	if (!catalog_sql_execute_once(&query))
+	{
+		/* errors have already been logged */
+		(void) semaphore_unlock(&(catalog->sema));
+		return false;
+	}
+
+	(void) semaphore_unlock(&(catalog->sema));
+
+	return true;
+}
+
+
+/*
+ * sentinel_update_replay_lsn updates our pgcopydb sentinel table end pos.
+ */
+bool
+sentinel_update_replay_lsn(DatabaseCatalog *catalog, uint64_t replay_lsn)
+{
+	sqlite3 *db = catalog->db;
+
+	if (db == NULL)
+	{
+		log_error("BUG: sentinel_update_replay_lsn: db is NULL");
+		return false;
+	}
+
+	char *sql = "update sentinel set replay_lsn = $1 where id = 1";
+
+	if (!semaphore_lock(&(catalog->sema)))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	SQLiteQuery query = { .errorOnZeroRows = true };
+
+	if (!catalog_sql_prepare(db, sql, &query))
+	{
+		/* errors have already been logged */
+		(void) semaphore_unlock(&(catalog->sema));
+		return false;
+	}
+
+	char replayLSN[PG_LSN_MAXLENGTH] = { 0 };
+	sformat(replayLSN, sizeof(replayLSN), "%X/%X", LSN_FORMAT_ARGS(replay_lsn));
+
+	/* bind our parameters now */
+	BindParam params[] = {
+		{ BIND_PARAMETER_TYPE_TEXT, "replay_lsn", 0, (char *) replayLSN }
+	};
+
+	int count = sizeof(params) / sizeof(params[0]);
+
+	if (!catalog_sql_bind(&query, params, count))
+	{
+		/* errors have already been logged */
+		(void) semaphore_unlock(&(catalog->sema));
+		return false;
+	}
+
+	/* now execute the query, which does not return any row */
+	if (!catalog_sql_execute_once(&query))
+	{
+		/* errors have already been logged */
+		(void) semaphore_unlock(&(catalog->sema));
+		return false;
+	}
+
+	(void) semaphore_unlock(&(catalog->sema));
+
+	return true;
+}
+
+
+/*
+ * sentinel_get fetches the current sentinel values
+ */
+bool
+sentinel_get(DatabaseCatalog *catalog, CopyDBSentinel *sentinel)
+{
+	sqlite3 *db = catalog->db;
+
+	if (db == NULL)
+	{
+		log_error("BUG: sentinel_get: db is NULL");
+		return false;
+	}
+
+	char *sql =
+		"select startpos, endpos, apply, write_lsn, flush_lsn, replay_lsn "
+		"  from sentinel "
+		" where id = 1";
+
+	if (!semaphore_lock(&(catalog->sema)))
```

Comment:
Instead of explicit serialisation, can't we rely on the default sqlite behavior while accessing the DB from multiple processes?

According to the https://www.sqlite.org/faq.html, sqlite would return SQLITE_BUSY when the the DB file is already opened by the other process, maybe we could retry few a times on SQLITE_BUSY and abort? WDYT?

Reply:
That's a good idea and that's what the code used to be doing up to recently. In a single-writer multiple-readers scenario, that'd be all we need for happily benefiting from SQLite concurrency facilities. Unfortunately our code is not single-writer, and we need to discipline it in a way that it is: that's why we have explicit serialization. It's up to pgcopydb code to ensure reliable operations by having a single SQLite writer at all times.

At least I failed to make it work any other way, and I tried.